edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/DeviceScopeIdentitiesCache.cs

```diff
@@ -0,0 +1,294 @@
+                catch (Exception e)
+                {
+                    Events.ErrorInRefreshCycle(e);
+                }
+
+                Events.DoneRefreshCycle(this.refreshRate);
+                await Task.Delay(this.refreshRate);
+            }
+        }
+
+        public async Task RefreshServiceIdentity(string deviceId)
+        {
+            try
+            {
+                Option<ServiceIdentity> serviceIdentity = await this.serviceProxy.GetServiceIdentity(deviceId);
+                await serviceIdentity
+                    .Map(s => this.HandleNewServiceIdentity(s))
+                    .GetOrElse(() => this.HandleNoServiceIdentity(deviceId));
+            }
+            catch (Exception e)
+            {
+                Events.ErrorRefreshingCache(e, deviceId);
+            }
+        }
+
+        public async Task RefreshServiceIdentity(string deviceId, string moduleId)
+        {
+            try
+            {
+                Option<ServiceIdentity> serviceIdentity = await this.serviceProxy.GetServiceIdentity(deviceId, moduleId);
+                await serviceIdentity
+                    .Map(s => this.HandleNewServiceIdentity(s))
+                    .GetOrElse(() => this.HandleNoServiceIdentity($"{deviceId}/{moduleId}"));
+            }
+            catch (Exception e)
+            {
+                Events.ErrorRefreshingCache(e, $"{deviceId}/{moduleId}");
+            }
+        }
+
+        public async Task RefreshServiceIdentities(IEnumerable<string> deviceIds)
+        {
+            List<string> deviceIdsList = Preconditions.CheckNotNull(deviceIds, nameof(deviceIds)).ToList();
+            foreach (string deviceId in deviceIdsList)
+            {
+                await this.RefreshServiceIdentity(deviceId);
+            }
+        }
+
+        public Task<Option<ServiceIdentity>> GetServiceIdentity(string deviceId, string moduleId)
+        {
+            Preconditions.CheckNonWhiteSpace(deviceId, nameof(deviceId));
+            Preconditions.CheckNonWhiteSpace(moduleId, nameof(moduleId));
+            return this.GetServiceIdentity($"{deviceId}/{moduleId}");
+        }
+
+        public async Task<Option<ServiceIdentity>> GetServiceIdentity(string id)
+        {
+            Preconditions.CheckNonWhiteSpace(id, nameof(id));
+            using (await this.cacheLock.LockAsync())
+            {
+                return this.serviceIdentityCache.TryGetValue(id, out StoredServiceIdentity storedServiceIdentity)
+                    ? storedServiceIdentity.ServiceIdentity
+                    : Option.None<ServiceIdentity>();
+            }
+        }
+
+        async Task HandleNoServiceIdentity(string id)
+        {
+            using (await this.cacheLock.LockAsync())
+            {
+                var storedServiceIdentity = new StoredServiceIdentity(id);
+                this.serviceIdentityCache[id] = storedServiceIdentity;
+                await this.SaveServiceIdentityToStore(id, storedServiceIdentity);
```

Comment:
This throws if id is not in the cache, isn't it possible that the id is not found?

Reply:
Not sure I understand.. where would this throw?